src/main/java/com/allocadia/carbonite/CarbonitePersist.java

```diff
@@ -0,0 +1,54 @@
+package com.allocadia.carbonite;
+
+import org.springframework.jdbc.core.support.JdbcDaoSupport;
+
+import com.allocadia.carbonite.utils.QueryUtils;
+import com.google.common.base.Joiner;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import lombok.SneakyThrows;
+
+public class CarbonitePersist<T extends PersistedObject> extends JdbcDaoSupport {
+
+    public Collection<T> save(Carbonite carbonite, Collection<T> objects) {
+        ModifiedObjects modifiedObjects = ModifiedObjects.from(objects);
+        Class<?> clazz = objects.iterator().next().getClass();
+        PersistenceInfo<PersistedObject> info = (PersistenceInfo<PersistedObject>) carbonite.getPersistenceInfo(clazz);
+        
+        persistNew(modifiedObjects.getNewObjects(), info);
+        //TODO: persistDirty
+    }
+
+    @SneakyThrows
+    private void persistNew(List<PersistedObject> newObjects, PersistenceInfo<PersistedObject> info) {
+        Map<Field, String> field2Column = info.getField2Column();
+        String baseSql = Joiner.on("\n").join(
+            "INSERT INTO " + info.getTableName(),
+            "(" + String.join(", ", field2Column.values()) + ")",
+            "VALUES",
+            "(" + QueryUtils.createParamList(field2Column.values().size()) + ")");
+        
+        List<Object[]> argList = new ArrayList<>();
+        for (PersistedObject persistedObject : newObjects) {
+            Object[] objectArgs = new Object[field2Column.keySet().size()];
+            
+            Iterator<Field> fieldIterator = field2Column.keySet().iterator();
+            int position = 0;
+            while (fieldIterator.hasNext()) {
+                objectArgs[position++] = fieldIterator.next().get(persistedObject);
+            }
+            
+            argList.add(objectArgs);
+        }
+        
+        super.getJdbcTemplate().batchUpdate(baseSql, argList);
```

Comment:
Fetching the newest N rows also assumes that the natural ordering is the same as the insertion order. And fetching the last N rows is probably also pretty slow.

I think you have to insert each row one at a time, and fetch the ID after each insert?

Here's the plain mysql way: https://dev.mysql.com/doc/refman/5.7/en/information-functions.html#function_last-insert-id

And spring seems to have a wrapper way that might be a little more generic: http://docs.spring.io/spring/docs/4.3.x/spring-framework-reference/html/jdbc.html#jdbc-auto-genereted-keys